### PR TITLE
Use library-go ApplyDeployment and ApplyDaemonset

### PR DESCRIPTION
**Pull request by mkenigs**
Use library-go ApplyDeployment and ApplyDaemonset

Our current functions in ./lib are out of date and we should be using
library-go instead

The library-go functions require passing the generation of the
previously applied deployment/daemonset, and this should be stored on an
MCO CRD. Because this currently doesn't exist, the existing generation
is passed as generation. This is no worse than how MCO's current
functions work, so the library-go functions should still be used until
an MCO CRD is created.

Files with the now unused functions and unused helper functions were
removed:
resourceapply/apps.go
resourcemerge/{apps.go,apps_test.go,core.go,core_test.go}

Helps https://github.com/openshift/machine-config-operator/issues/819
Depends https://github.com/openshift/machine-config-operator/pull/2833

### CODE REVIEW

**Comment by mkenigs** [comment]
@deads2k does this look right? Just the last commit

File: pkg/operator/sync.go [hunk 1]
```diff
@@ -514,11 +514,12 @@ func (optr *Operator) applyManifests(config *renderConfig, paths manifestPaths)
 			return err
 		}
 		d := resourceread.ReadDaemonSetV1OrDie(dBytes)
-		_, updated, err := mcoResourceApply.ApplyDaemonSet(optr.kubeClient.AppsV1(), d)
+		daemonset, updated, err := resourceapply.ApplyDaemonSet(context.TODO(), optr.kubeClient.AppsV1(), optr.libgoRecorder, d, optr.daemonsetGeneration)
 		if err != nil {
 			return err
 		}
 		if updated {
+			optr.daemonsetGeneration = daemonset.Generation
```
Comment (deads2k) [inline pkg/operator/sync.go:522]: does this get persisted so when you restart you have it?

File: pkg/operator/sync.go [hunk 2]
```diff
@@ -553,11 +554,12 @@ func (optr *Operator) syncMachineConfigController(config *renderConfig) error {
 	}
 	mcc := resourceread.ReadDeploymentV1OrDie(mccBytes)
 
-	_, updated, err := mcoResourceApply.ApplyDeployment(optr.kubeClient.AppsV1(), mcc)
+	deployment, updated, err := resourceapply.ApplyDeployment(context.TODO(), optr.kubeClient.AppsV1(), optr.libgoRecorder, mcc, optr.deploymentGeneration)
 	if err != nil {
 		return err
 	}
 	if updated {
+		optr.deploymentGeneration = deployment.Generation
```
Comment (deads2k) [inline pkg/operator/sync.go:562]: does this get persisted so when you restart you have it?

**Comment by deads2k** [comment]
just the question about persistence to ensure that the right happens on restart.  Looks good otherwise.

**Comment by deads2k** [comment]
example from openshiftapiserver status

```
status
  generations:
  - group: apps
    hash: ""
    lastGeneration: 5
    name: apiserver
    namespace: openshift-apiserver
    resource: deployments
```

**Comment by yuqi-zhang** [comment]
@mkenigs and I did some investigating on storing generations properly to use for apply functions, and this is what we've determined (please correct any mistakes):

Using OpenshiftAPIServer object as an example (i.e. `oc describe OpenShiftAPIServer/cluster`), generations is supposed to be stored on some object the operator manages.

The first issue here is, the MCO never had any object tracking daemonset/deployment object generations like this.

The second issue is, the MCO doesn't actually have an equivalent object to track with. The MCO creates the following CRDS:

- Containerruntimeconfig/Kubeletconfig

- MachineConfigs

- MachineConfigPools

- ControllerConfig

Of these CRDs we don't have a direct equivalent of what OpenshiftAPIServer is in the above context. The closest is ControllerConfig which is a cluster level config but only used by the MCO (mayhaps naming is not very apt), and is used by the operator pod to read cluster objects (e.g. proxy, osimage) and populate for the machine-config-controller to render into templates for the machine-config-daemons.

So when that in mind, our options seem to be:

1. For now, store the generations we last acted on in the operator Go struct. This is the least optimal since every time the pod restarts, this would get reset which triggers a forced apply (which shouldn't cause any issues otherwise), but normal syncs would not trigger applies
2. Every time we want to apply, read the observedGeneration of the object first, because as far as I understand that is basically what the library-go functions are asking for: the last generation we acted upon. If this is an ok path, we wouldn't have to rework any object to store generations data
3. Combine 1 and 2 so we read on pod start and then sync via operator. This saves a read operations every sync
4. Add daemonset/deployment data to ControllerConfig and store generations properly there, although it is not what the object was originally created to do
5. Create a new CRD (OpenshiftMachineConfig...?) that for now, is just doing the generations part of the equivalent of OpenShiftAPIServer

Not sure which approach is best for our case.

**Comment by cgwalters** [comment]
> The closest is ControllerConfig which is a cluster level config but only used by the MCO (mayhaps naming is not very apt), and is used by the operator pod to read cluster objects (e.g. proxy, osimage) and populate for the machine-config-controller to render into templates for the machine-config-daemons.

Tangentially related, I keep thinking of `ControllerConfig` as basically a private channel between the operator pod and the controller pod.  Which could go away entirely if they were in the same process.

**Comment by deads2k** [comment]
> * For now, store the generations we last acted on in the operator Go struct. This is the least optimal since every time the pod restarts, this would get reset which triggers a forced apply (which shouldn't cause any issues otherwise), but normal syncs would not trigger applies
> * Every time we want to apply, read the observedGeneration of the object first, because as far as I understand that is basically what the library-go functions are asking for: the last generation we acted upon. If this is an ok path, we wouldn't have to rework any object to store generations data
> * Combine 1 and 2 so we read on pod start and then sync via operator. This saves a read operations every sync
> * Add daemonset/deployment data to ControllerConfig and store generations properly there, although it is not what the object was originally created to do
> * Create a new CRD (OpenshiftMachineConfig...?) that for now, is just doing the generations part of the equivalent of OpenShiftAPIServer

I suggest creating a CRD to start the equivalent of openshiftapiserver.operator.openshift.io.  This would also provide a spot for operand loglevel, operator loglevel, unsupported overrides, detailed status to add additional degraded loops.

Kicking deployments and daemonsets on every process restart isn't a state we should ship.

**Comment by mkenigs** [comment]
/hold
Going to wait for https://issues.redhat.com/browse/MCO-163 to do this correctly

**Comment by kikisdeliveryservice** [comment]
> /hold Going to wait for https://issues.redhat.com/browse/MCO-163 to do this correctly

Note: I removed the BZ as this is going to be worked on via the above jira issue and so that the underlying BZ can be closed based on the other 2 PRs that merged.

**Pull request closed without merging**